tensorflow_quantum/datasets/spin_system.py

```diff
@@ -0,0 +1,240 @@
+        self.path = path
+
+    def __lt__(self, other):
+        return self.g < other.g
+
+    def hamiltonian(self) -> sps.csr_matrix:
+        """Load the hamiltonian into a scipy sparse CSR matrix."""
+        return sps.load_npz(self.path + '/H.npz')
+
+    def ground_state(self) -> np.ndarray:
+        """Load the ground state wave function into a complex numpy array."""
+        return np.load(self.path + '/groundstate.npy')
+
+    def final_state(self):
+        """Load the circuit wave function."""
+        return np.load(self.path + '/final_state.npy')
+
+
+def unique_name():
+    """
+    Generator to generate an infinite number of unique names.
+
+    Yields:
+        Strings of the form 'theta_<integer>'
+
+    """
+    num = 0
+    while True:
+        yield 'theta_' + str(num)
+        num += 1
+
+
+class SpinSystemDataset(object):
+    """Abstract class that handles a single instance of the system for a
+    single set of order parameters.
+    """
+
+    def __init__(self, nspins: int, system_name: str):
+        """Instantiate this data set.
+
+        Args:
+            nspins: The number of spins in the system.
+            system_name: Name of the system we want to load.
+        """
+
+        # TODO: How should downloading and storing the data be handled?
+        data_path = '/tmp/data/' + system_name + "_{}/".format(int(nspins))
+        self.nspins = nspins
+        if not os.path.exists(data_path):
+            os.mkdir(data_path)
+            self._download_dataset(data_path)
+        # TODO: This only works for a single order parameter. For
+        #  the XY model or XYZ model, this will have to be generalized.
+        self.systems = self._get_spin_systems(data_path)
+
+    def __str__(self):
+        # TODO: This only works for a single order parameter. For
+        #  the XY model or XYZ model, this will have to be generalized.
+        return "System with {} order parameters in range [{},{}]".format(
+            len(self), self.systems[0].g, self.systems[-1].g)
+
+    def __len__(self):
+        return len(self.systems)
+
+    def _get_spin_systems(self, path: str) -> List[SpinSystem]:
+        systems = []
+        for directory in os.listdir(path):
+            with open(path + directory + '/stats.txt', 'r') as file:
+                g = float(directory)
+                lines = file.readlines()
+                res_e = float(lines[0].split('=')[1].strip('\n'))
+                fidelity = float(lines[0].split('=')[1].strip('\n'))
+            gs_energy = np.load(path + directory + '/energy.npy')[0]
+            parameters = np.load(path + directory + '/params.npy')
+            systems.append(
+                SpinSystem(g, res_e, fidelity, gs_energy, parameters,
+                           path + directory))
+        return sorted(systems)
+
+    def _download_dataset(self, path: str):
+        # TODO: Implement this.
+        pass
+
+
+class TFIChain(SpinSystemDataset):
```

Comment:
Could we turn this into a function that returns a tuple of lists/arrays to the effect of ?
```python
def tfi_chain(...):
    .
    .
    .
    return circuits(with parameters resolved in), g, gs_energy, res_e
```
Where `circuits[i]` has corresponding information `g[i]`, `gs_energy[i]` or `res_e[i]`. This feels more natural for a user reading the documentation who could then do something like:
```python
circuits, g, gse, _ = tfq.datasets.tfi_chain(...)
data_tensor = tfq.convert_to_tensor(circuits)
label_tensor = tf.convert_to_tensor(gse)
```

Reply:
Doing this with a function should be fine.

But returning a circuit for each parameter, would'nt that lead to 80 copies of the same circuit? Or can we resolve multiple parameter sets for a single circuit? My reasoning was that it's best to create the circuit once, and then allow the user to feed the parameters in batches. Something like
```python
def tfi_chain(...):
    .
    .
    .
    return circuit, resolved_params, g, gs_energy, res_e
```